### ISSUE TITLE
Multi-output GPs

### ISSUE THREAD

**Issue by willtebbutt** [JuliaGaussianProcesses/AbstractGPs.jl#9]
There are a couple of options here.

## Multi-output GPs are just GPs

A multioutput GP can just be viewed as a single output GP on an extended input domain. Specifically, suppose we're interested in modelling functions $\mathcal{X} \to \mathbb{R}^N$. You can always rephrase this as a function $\mathcal{X} \times \{1, ..., N\} \to \mathbb{R}$. Consequently you can always re-write a multioutput GP as a single-output GP on an extended input domain.

## Multi-output GPs are there own special thing

The other approach is to introduce a custom object here that really treats a multi-output GP as a separate thing, and introduce new API features / conventions to handle them.

## Pros and Cons

There are a couple of advantages to the first approach: namely that there's no separate notion of a multi-output GP floating around the code base, just extra abstractions (probably just an extra `AbstractVector` type and some specifically "multi-output" kernels). This also means that things like pseudo-point approximations will just work with the multi-output models and be non-mean-field by default.

That being said, the second approach makes the multi-output structure _really_ explicit, and it's often the case that you want to be able to get at this for computational reasons. However, as I mentioned we can just introduce some new abstractions for multi-output GPs if we go with the everything-is-just-a-GP approach, making this look slightly less like an advantage.

Overall I definitely favour the everything-is-a-GP approach because it fits nicely with our current way of doing things, and makes multi-output GPs trivial from an API perspective -- if you're comfortable with things like the `ColVecs` and `RowVecs` abstractions, multi-output GPs will be completely natural.

**Comment by sharanry** [comment]
>A multioutput GP can just be viewed as a single output GP on an extended input domain. Specifically, suppose we're interested in modelling functions $\mathcal{X} \to \mathbb{R}^N$. You can always rephrase this as a function $\mathcal{X} \times {1, ..., N} \to \mathbb{R}$. Consequently you can always re-write a multioutput GP as a single-output GP on an extended input domain.

I am not sure how this is made possible. Are we assuming all multi-output functions we model are invertible in some sense and go on to the model the inverse of the function?

Few other ways we could design a multi-output GP.  
1. Have separate kernels for each output dimension and have combined kernel which would incorporate these separate kernels along with a co-variances between the two output dimensions. Refer to pg 7 in [1] OR have shared independent multi-output kernel as suggested in [2]. 
3. Shared pseudo-points for all dimensions OR separate set of pseudo points for each dimensions. Also discussed in [2].

There lot of other considerations made in [2]. I haven't gone through all of them in detail yet.

[1] - GPSS'17 - Multiple-output Gaussian processes [slides](http://gpss.cc/gpss17/slides/multipleOutputGPs.pdf)
[2] - A Framework for Interdomain and Multioutput Gaussian Processes [arxiv](https://arxiv.org/pdf/2003.01115.pdf)

**Comment by willtebbutt** [comment]
> I am not sure how this is made possible. Are we assuming all multi-output functions we model are invertible in some sense and go on to the model the inverse of the function?

No, nothing like this. It's just re-writing the function slightly. The construction that I discussed above is basically just a re-labelling -- instead of treating each element of the domain as mapping to a vector, you add an extra dimension to the domain that says which output you're mapping to.

More concretely, let `f(x)` return a `D`-dimensional vector of reals, and `g((x, d))` return a reals. If you require that they satisfy
```julia
f(x)[d] = g((x, d))
```
then hopefully it's clear what I meant by converting a vector-valued function to a scalar-valued function on an extended domain.

You can do the same for GPs. You can think of a kernel over a vector-valued GP as a binary function that returns a matrix (that will be positive definite), or you can think of it just as a kernel that on an extended domain.

Concretely, let `K(x, y)` return a `D x D` (positive definite) matrix of reals, and `k((x, p), (y, q))` return a real satisfying
```julia
K(x, y)[p, q] == k((x, p), (y, q))
```
So what we've done here is just re-write things such that the input to a GP also encodes which output we're talking about.

The extended-domain approach is strictly more general than the matrix-valued-kernel approach.

> I found a few other ways we could model a multi-output GP or rather the choices we need to make while modelling a multi-output GP:

The benefit of extended-domain view of the world is that these aren't choices that you have to make.

It's worth noting that all of the inter-domain stuff is subsumed by the GPPP framework in Stheno -they're kind of trivial once you've broken everything down. It's then just a question of whether you can express the generative model in Stheno required to do the interdomain stuff -- the limiting factor being whether or not something has been implemented.

**Comment by sharanry** [comment]
@willtebbutt I understand what you meant now. Thanks!

I want to clarify a few things,
>```julia
>f(x)[d] = g((x, d))
>K(x, y)[p, q] == k((x, p), (y, q))
>```
1) Are `d`, `p` and `q` here are integers/indices?
2) If so, treating these as just another input(through an extended domain) seems odd to me as they are discrete quantities. To confirm, is `k` in `k((x, p), (y, q))` just a regular kernel which we already have in `KernelFunctions.jl`?

**Comment by willtebbutt** [comment]
1. Yup
2. You probably wouldn't use one of the kernels that we've already got in `KernelFunctions.jl` in this way, the point is more that you can in principle write things down this way, and build input types on the back of this that appeal to this mathematical structure.

### For Example

So, for example, suppose you've got a two-output process `f` (perhaps a co-regionalisation model), and we'll call the first output `f_1` and the second `f_2`, and you want the `FiniteGP` object that corresponds to the multivariate Normal of dimensionality `2N` that you get by "stacking" `f_1(x)` on top of `f_2(x)` for some vector of inputs `x` of length `N`.

Suppose the kernel of `f` literally accepts inputs of the form `(x_n, p_n)` where `x_n` is the nth input location and `p_n` the process to with the nth input corresponds (either 1 or 2 in our case). If you wanted to do something literal (but inefficient) you could construct the length `2N` vector
```julia
x_naive = [(x_1, 1), (x_2, 1), ..., (x_N, 1), (x_1, 2), (x_2, 2), ..., (x_N, 2)]
```
and construct `f(x_naive)`. Provided that you've got all of your fallbacks implemented correctly, this should work.

Alternatively, you could introduce an input type `MultiVector` (that's a horrible name, and we shouldn't adopt it) with indexing defined such that
```julia
x_multi = MultiVector(x, 2)
x_naive .== x_multi
```
Then when you construct
```julia
f(x_multi)
```
it can look at the input type and know what computation it is that you want to do, without actually having to augment all of your inputs etc.

### Isn't this a little convoluted?

This approach is probably a bit more complicated in the simplest cases. For example, when each output is fully observed, a special vector-valued GP _might_ be slightly simpler, but I don't believe the gain is that much. Certainly nothing is gained / lost computationally.

Moreover you don't have to have to consider especially complicated situations before things become hard to do with a vector-valued GP. Consider the case in which most of the time only 1 or 2 outputs are observed at each input. You would need to introduce some new data structure to efficiently represent your matrices of outputs sparsely (I'm assuming that for a special vector-valued GP object, `y` will typically be a `Matrix`). The standard sparse matrices are unlikely to be performant in these cases as they're known to have quite high overhead. Certainly you might want a special input type in the above framework, but the point is that the vector-valued GP doesn't win in this situation.

The main advantage, from my perspective, to the proposed approach is the lack of a need to extend the GP interface and introduce a special new GP object. Certainly you would need some new abstractions, but they would be need abstractions at a level that we're already comfortable with -- for example we use `ColVecs` all over the place and implement kernels that know how to handle that. This approach is just more of the same, so once you've got your head around `ColVecs` and how that keeps in the interface simple, you should be happy with the kinds of things you might do to get a multi-output GP object to work.

**Comment by theogf** [comment]
I think I finally understood your approach, which was not intuitive for me. But I agree this is probably the most general approach. From there I see two important points to be taken :
- Write a very clear explanation in the docs of how the multi-output GPs are defined (since it's quite different then how most models are explained)
- Be sure to enforce special cases smartly (if the outputs are considered independent one should store the kernel matrix efficiently)

**Comment by sharanry** [comment]
@willtebbutt and I had a good discussion on design of input datatype for multi-output GPs. I have some doubts regarding the necessity for the datatype to behave in the way described in this [comment](https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/9#issuecomment-637659488) above. I will try making a case for the alternate approach here.

We aim to model data with input vectors like `x = [ x_1, x_2, ..., x_n ]` of dimension `n` and output vectors like `y = [ y_1, y_2, ... y_k]` of dimension `k`.

I propose we create a datatype called `MultiGPInput` which accepts input vector `x` and output dimension `k` (same as what WIll had proposed) to **emulate** a vector of `Real`s of length `n*k`,

`x_multi = [ x_1, x_2, ..., x_n, x_1, x_2, ..., x_n, ... , x_1, x_2, ..., x_n]` where `x` is repeated `k` times. 

The reason I propose this as opposed to using a vector of tuples suggested above is that if feel the second element of the tuple denoting the dimension of the output is unecessary. With the proposed scheme, this is implicitly available from the position of the element accessed from `x_multi`.

If we were to support prediction on incomplete test input (i.e, when all observations are not available at test time), we could maintain null values in those places and handle those null values within the multi-output kernels.

 @theogf @devmotion I would love to hear you thoughts on this. :)

**Comment by theogf** [comment]
@sharanry About multi-output I feel that what is not dealt with right now is that what if one also has multiple input sources, how would that work in this case?

**Comment by sharanry** [comment]
> @sharanry About multi-output I feel that what is not dealt with right now is that what if one also has multiple input sources, how would that work in this case?

Here `n` is the no. of dimensions of the input space and `x` is one instance of the input. Aren't multi dimensional inputs dealt with by the existing kernels?
For example, 
```julia
julia> k = SqExponentialKernel()
Squared Exponential Kernel

julia> k(rand(3), rand(3))
0.7218790877773378

```
or have I understood your question wrong?

**Comment by theogf** [comment]
Yeah I am sorry that was not clear, but I think I got confused myself. I did not mean the multivariate case, but more that you could have two time series (where inputs are collected at the same time), with different inputs domains and you want to model the output of both at the same time. But as I wrote it, it becomes clear that this is a very specific case, and that it can be handled easily by simply concatenating the inputs and having an adapted kernel.

**Comment by willtebbutt** [comment]
The reason that I don't feel that this proposal doesn't work properly is that the `AbstractVector` interpretation of a data structure formatting in the way that you propose would essentially have repeated inputs if you were to call `getindex`. e.g.
```julia
x = MultiGPInput(randn(3), 5) # length 15 vector
x[1] == x[4] # returns true
```
which is problematic from the perspective of this data type making sense in isolation (which is a thing that I really think that it should do).

The benefit of at least _thinking_ about each entry as a Tuple is that the above issue doesn't happen, and the vector makes sense by itself.

> with different inputs domains and you want to model the output of both at the same time.

@theogf are you thinking about having multiple bits of side information or something here? i.e. inputs other than time?

**Comment by willtebbutt** [comment]
Closing this because things have progressed significantly since this discussion.
